### PR TITLE
Refactor `Colorize::Mode` enum

### PR DESCRIPTION
**Pull request by straight-shoota**
Extracted from #7690

This is mostly an internal refactoring. Only effect on the API is that invalid symbols are now a compile time error instead of runtime error. That could be considered a breaking change, but it's an improvement for error detection.

### CODE REVIEW

File: src/colorize.cr [hunk 1]
```diff
@@ -226,36 +226,52 @@ module Colorize
...
+    # Underlines the text.
+    Underline
+    # Makes the text blink slowly.
+    Blink
+    # Swaps the foreground and background colors of the text.
+    Reverse
+    # Makes the text invisible.
+    Hidden
+
+    def code
```
Comment (Sija) [inline src/colorize.cr:251]:
```suggestion
    def code : Char
```

File: src/colorize.cr [hunk 2]
```diff
@@ -226,36 +226,52 @@ module Colorize
...
+      case self
```
Comment (Sija) [inline src/colorize.cr:252]: Couldn't this be turned into an exhaustive enum?

Comment (straight-shoota) [reply src/colorize.cr:252]: Exhaustive case doesn't work for flag enums.

File: src/colorize.cr [hunk 3]
```diff
@@ -308,15 +317,10 @@ struct Colorize::Object(T)
...
-    {% for name in MODES %}
-      if mode == :{{name.id}}
-        @mode |= MODE_{{name.upcase.id}}_FLAG
-        return self
-      end
-    {% end %}
-
-    raise ArgumentError.new "Unknown mode: #{mode}"
+  # Adds *mode* to the text's decorations.
+  def mode(mode : Mode)
```
Comment (j8r) [inline src/colorize.cr:321]: The `self` return type has been removed.

File: src/colorize.cr [hunk 4]
```diff
@@ -379,25 +383,24 @@ struct Colorize::Object(T)
     last_color_is_default =
       @@last_color[:fore] == ColorANSI::Default &&
         @@last_color[:back] == ColorANSI::Default &&
-        @@last_color[:mode] == 0
+        @@last_color[:mode] == Mode::None
```
Comment (caspiano) [inline src/colorize.cr:386]:
```suggestion
        @@last_color[:mode].none?
```

File: src/colorize.cr [hunk 5]
```diff
@@ -413,15 +416,22 @@ struct Colorize::Object(T)
...
+        mode.each do |flag|
+          # Enum#each yields each member flag. Bright and bold have the same value
+          # and would show up as duplicate, so we need to handle this special case.
+          if flag.bright?
+            if printed_bright
+              next
+            else
+              printed_bright = true
+            end
           end
```
Comment (oprypin) [inline src/colorize.cr:430]: Consider this refactor: https://github.com/oprypin/crystal/commit/5956803b29c408a64f9ac3c00810151b57e2d047

File: src/colorize.cr (hunk 2 above)
Comment (oprypin) [reply src/colorize.cr:243]: I think it just highlights that it's a strange thing to do - picking out results for individual values in a flags enum. It has a tendency of misunderstanding - for an actual combined flags value the other sub-values will be ignored.
Also consider not making this method public.
I address this in https://github.com/oprypin/crystal/commit/5956803b29c408a64f9ac3c00810151b57e2d047

**Review by oprypin — APPROVED** [review]

File: src/colorize.cr [hunk 6]
```diff
@@ -253,37 +244,50 @@ module Colorize
...
+    Blink
+    # Swaps the foreground and background colors of the text.
+    Reverse
+    # Makes the text invisible.
+    Hidden
+  end
+end
+
+private def each_code(mode : Colorize::Mode)
+  yield '1' if mode.bold?
```
Comment (vlazar) [inline src/colorize.cr:272]: I know the result would be the same, but for readability I wish it was explicitly covering all enum members.
```cr
yield '1' if mode.bold? || mode.bright?
```

Maybe compiler could be (is) smart enough to make it noop, but humans would see all cases covered.
Or maybe just comment.

Comment (straight-shoota) [reply src/colorize.cr:272]: I don't think the compiler can easily optimize this.

I'll add a comment to the member definition. That's probably more useful than here in this internal implementation.

**Pull request merged**